backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+	if url == "" {
+		return nil, errors.New("forbidden to download - check sharing permission")
+	}
+	req, err := http.NewRequestWithContext(ctx, method, url, nil)
+	if err != nil {
+		return nil, err
+	}
+	fs.FixRangeOption(options, o.size)
+	fs.OpenOptionAddHTTPHeaders(req.Header, options)
+	if o.size == 0 {
+		// Don't supply range requests for 0 length objects as they always fail
+		delete(req.Header, "Range")
+	}
+	err = o.fs.pacer.Call(func() (bool, error) {
+		res, err = o.fs.client.Do(req)
+		return o.fs.shouldRetry(ctx, res, err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// open a url for reading
+func (o *Object) open(ctx context.Context, url string, options ...fs.OpenOption) (in io.ReadCloser, err error) {
+	res, err := o.httpResponse(ctx, url, "GET", options)
+	if err != nil {
+		return nil, fmt.Errorf("open file failed: %w", err)
+	}
+	return res.Body, nil
+}
+
+// Open an object for read
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (in io.ReadCloser, err error) {
+	if o.id == "" {
+		return nil, errors.New("can't download - no id")
+	}
+	info, err := o.fs.getFile(ctx, o.id)
+	if err != nil {
+		return nil, fmt.Errorf("can't fetch download link: %w", err)
+	}
+	if err = o.setMetaData(info); err != nil {
+		return nil, err
+	}
+	if o.link == nil {
+		if o.size == 0 {
+			// zero-byte objects may have no download link
+			return io.NopCloser(bytes.NewBuffer([]byte(nil))), nil
+		}
+		return nil, errors.New("can't download - no link to download")
+	}
+	return o.open(ctx, o.link.URL, options...)
+}
+
+// Update the object with the contents of the io.Reader, modTime and size
+//
+// If existing is set then it updates the object rather than creating a new one.
+//
+// The new object may have been created if an error is returned
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+	size := src.Size()
+	remote := o.Remote()
+
+	// Create the directory for the object if it doesn't exist
+	leaf, dirID, err := o.fs.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return err
+	}
+
+	// We have to fall back to upload + rename
+	tempName := "rcloneTemp" + random.String(8)
+	info, err := o.fs.upload(ctx, in, tempName, dirID, size, options...)
+	if err != nil {
+		return err
+	}
+
+	// upload was successful, need to delete old object before rename
+	if err = o.Remove(ctx); err != nil {
+		return fmt.Errorf("failed to remove old object: %w", err)
+	}
+
+	// rename also updates metadata
+	if info, err = o.fs.renameObject(ctx, info.ID, leaf); err != nil {
+		return fmt.Errorf("failed to rename temp object: %w", err)
+	}
+	return o.setMetaData(info)
+}
+
+// Check the interfaces are satisfied
+var (
+	// _ fs.ListRer         = (*Fs)(nil)
+	// _ fs.ChangeNotifier  = (*Fs)(nil)
+	// _ fs.PutStreamer     = (*Fs)(nil)
```

Comment:
Not having PutStream is sad - this means that `rclone mount` will need `--vfs-cache-mode writes` or `full` in order to upload files amongst other things.

Reply:
Yes `rclone rcat` is a good way of testing. Note that it will stream larger files to disk unless you advertise `PutStream`.